modules/flist/flist.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zosv2/modules"
+)
+
+const (
+	defaultStorage = "zdb://hub.grid.tf:9900"
+	defaultRoot    = "/var/modules/flist"
+)
+
+type flistModule struct {
+	// root directory where all
+	// the working file of the module will be located
+	root string
+
+	// underneath are the path for each
+	// sub folder used by the flist module
+	flist      string
+	cache      string
+	mountpoint string
+	pid        string
+	log        string
+}
+
+// New creates a new flistModule
+func New(root string) modules.Flister {
+	if root == "" {
+		root = defaultRoot
+	}
+
+	// prepare directory layout for the module
+	for _, path := range []string{"flist", "cache", "mountpoint", "pid", "log"} {
+		if err := os.MkdirAll(filepath.Join(root, path), 0770); err != nil {
+			panic(err)
+		}
+	}
+
+	return &flistModule{
+		root:       root,
+		flist:      filepath.Join(root, "flist"),
+		cache:      filepath.Join(root, "cache"),
+		mountpoint: filepath.Join(root, "mountpoint"),
+		pid:        filepath.Join(root, "pid"),
+		log:        filepath.Join(root, "log"),
+	}
+}
+
+// Mount implements the Flister.Mount interface
+func (f *flistModule) Mount(url, storage string) (string, error) {
+	sublog := log.With().Str("url", url).Str("storage", storage).Logger()
+	sublog.Info().Msg("request to mount flist")
+
+	if storage == "" {
+		storage = defaultStorage
+	}
+
+	rc, err := downloadFlist(url)
```

Comment:
Check how 0-core does it, because we also now cache the flists as well. so we don't download the same flist over and over.

Reply:
Indeed, that is the plan, I still have a todo in the downloadFlist function for that.